File: modules/debugger/src/main/java/com/falsepattern/zigbrains/debugger/ZigDefaultDebuggerDriverConfigurationProvider.java
package com.falsepattern.zigbrains.debugger;

import com.falsepattern.zigbrains.ZigBundle;
import com.falsepattern.zigbrains.debugbridge.ZigDebuggerDriverConfigurationProvider;
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings;
import com.falsepattern.zigbrains.debugger.toolchain.DebuggerAvailability;
import com.falsepattern.zigbrains.debugger.toolchain.DebuggerAvailability.GDBBinaries;
import com.falsepattern.zigbrains.debugger.toolchain.DebuggerAvailability.LLDBBinaries;
import com.falsepattern.zigbrains.debugger.toolchain.DebuggerAvailability.MSVCBinaries;
import com.falsepattern.zigbrains.debugger.toolchain.DebuggerKind;
import com.falsepattern.zigbrains.debugger.toolchain.ZigDebuggerToolchainService;
import com.falsepattern.zigbrains.debugger.win.MSVCDriverConfiguration;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DoNotAskOption;
import com.intellij.openapi.ui.MessageDialogBuilder;
import com.intellij.openapi.ui.Messages;
import com.jetbrains.cidr.ArchitectureType;
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration;
import com.jetbrains.cidr.execution.debugger.backend.gdb.GDBDriverConfiguration;
import com.jetbrains.cidr.execution.debugger.backend.lldb.LLDBDriverConfiguration;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.nio.file.Path;
import java.util.function.Supplier;

public class ZigDefaultDebuggerDriverConfigurationProvider implements ZigDebuggerDriverConfigurationProvider {
    private static boolean availabilityCheck(Project project, DebuggerKind kind) {
        val service = ZigDebuggerToolchainService.getInstance();
        val availability = service.debuggerAvailability(kind);

        final String message, action;
        switch (availability.kind()) {
            case Bundled, Binaries -> {
                return true;
            }
            case Unavailable -> {
                return false;
            }
            case NeedToDownload -> {
                message = "Debugger is not loaded yet";
                action = "Download";
            }
            case NeedToUpdate -> {
                message = " Debugger is outdated";
                action = "Update";
            }
            default -> throw new AssertionError("This should never happen");
        }

        val downloadDebugger = ZigDebuggerSettings.getInstance().downloadAutomatically || showDialog(project, message, action);

        if (downloadDebugger) {
            val result = ZigDebuggerToolchainService.getInstance().downloadDebugger(project, kind);
            if (result instanceof ZigDebuggerToolchainService.DownloadResult.Ok)
                return true;
        }
        return false;
    }

    private static boolean showDialog(Project project, String message, String action) {
        val doNotAsk = new DoNotAskOption.Adapter() {
            @Override
            public void rememberChoice(boolean isSelected, int exitCode) {
                if (exitCode == Messages.OK) {
                    ZigDebuggerSettings.getInstance().downloadAutomatically = isSelected;
                }
            }
        };

        return MessageDialogBuilder.okCancel(ZigBundle.message("unable.to.run.debugger"), message)
                                   .yesText(action)
                                   .icon(Messages.getErrorIcon())
                                   .doNotAsk(doNotAsk)
                                   .ask(project);
    }

    @Override
    public @Nullable Supplier<DebuggerDriverConfiguration> getDebuggerConfiguration(Project project, boolean isElevated, boolean emulateTerminal) {
        val settings = ZigDebuggerSettings.getInstance();
        val service = ZigDebuggerToolchainService.getInstance();
        val kind = settings.debuggerKind;
        if (!availabilityCheck(project, kind)) {
            return null;
        }
        val availability = service.debuggerAvailability(kind);
        return switch (availability.kind()) {
            case Bundled -> () -> (switch (kind) {
                case LLDB -> new ZigLLDBDriverConfiguration(isElevated, emulateTerminal);
                case GDB -> new ZigGDBDriverConfiguration(isElevated, emulateTerminal);
                case MSVC -> throw new AssertionError("MSVC is never bundled");
            });
            case Binaries -> {
                val bin = (DebuggerAvailability.Binaries) availability;
                yield () -> (switch (bin.binariesKind()) {
                    case LLDB -> new ZigCustomBinariesLLDBDriverConfiguration((LLDBBinaries) bin, isElevated, emulateTerminal);
                    case GDB -> new ZigCustomBinariesGDBDriverConfiguration((GDBBinaries) bin, isElevated, emulateTerminal);
                    case MSVC -> new ZigMSVCDriverConfiguration((MSVCBinaries) bin, isElevated, emulateTerminal);
                });
            }
            case Unavailable, NeedToDownload, NeedToUpdate -> throw new AssertionError("This should never happen");
        };
    }

    //region GDB
    @RequiredArgsConstructor
    public static class ZigGDBDriverConfiguration extends GDBDriverConfiguration {
        private final boolean isElevated;
        private final boolean emulateTerminal;

        @Override
        public @NotNull String getDriverName() {
            return "Zig GDB";
        }

        // TODO: investigate attach to process feature separately
        @Override
        public boolean isAttachSupported() {
            return false;
        }

        @Override
        public boolean isElevated() {
            return isElevated;
        }

        @Override
        public boolean emulateTerminal() {
            return emulateTerminal;
        }
    }

    private static class ZigCustomBinariesGDBDriverConfiguration extends ZigGDBDriverConfiguration {
        private final GDBBinaries binaries;
        public ZigCustomBinariesGDBDriverConfiguration(GDBBinaries binaries, boolean isElevated, boolean emulateTerminal) {
            super(isElevated, emulateTerminal);
            this.binaries = binaries;
        }

        @Override
        protected @NotNull String getGDBExecutablePath() {
            return binaries.gdbFile().toString();
        }
    }
    //endregion GDB

    //region LLDB
    @RequiredArgsConstructor
    public static class ZigLLDBDriverConfiguration extends LLDBDriverConfiguration {
        private final boolean isElevated;
        private final boolean emulateTerminal;

        @Override
        public @NotNull String getDriverName() {
            return "Zig LLDB";
        }

        @Override
        public boolean isElevated() {
            return isElevated;
        }

        @Override
        public boolean emulateTerminal() {
            return emulateTerminal;
        }
    }

    private static class ZigCustomBinariesLLDBDriverConfiguration extends ZigLLDBDriverConfiguration {
        private final LLDBBinaries binaries;
        public ZigCustomBinariesLLDBDriverConfiguration(LLDBBinaries binaries, boolean isElevated, boolean emulateTerminal) {
            super(isElevated, emulateTerminal);
            this.binaries = binaries;
        }

        @Override
        public boolean useSTLRenderers() {
            return false;
        }

        @Override
        protected @NotNull File getLLDBFrameworkFile(@NotNull ArchitectureType architectureType) {
            return binaries.frameworkFile().toFile();
        }

        @Override
        protected @NotNull File getLLDBFrontendFile(@NotNull ArchitectureType architectureType) {
            return binaries.frontendFile().toFile();
        }
    }
    //endregion LLDB

    //region MSVC
    @RequiredArgsConstructor
    public static class ZigMSVCDriverConfiguration extends MSVCDriverConfiguration {
        private final MSVCBinaries binaries;
        private final boolean isElevated;
        private final boolean emulateTerminal;

        @Override
        public @NotNull String getDriverName() {
            return "Zig MSVC";
        }

        @Override
        protected Path getDebuggerExecutable() {
            return binaries.msvcFile();
        }

        @Override
        public boolean isElevated() {
            return isElevated;
        }

        @Override
        public boolean emulateTerminal() {
            return emulateTerminal;
        }
    }

    //endregion MSVC
}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/debugger/ZigDefaultDebuggerDriverConfigurationProvider.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.debugger

import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings
import com.falsepattern.zigbrains.debugger.toolchain.*
import com.falsepattern.zigbrains.debugger.win.MSVCDriverConfiguration
import com.falsepattern.zigbrains.shared.coroutine.withEDTContext
import com.falsepattern.zigbrains.shared.sanitizedPathString
import com.falsepattern.zigbrains.zig.ZigLanguage
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DoNotAskOption
import com.intellij.openapi.ui.MessageDialogBuilder
import com.intellij.openapi.ui.Messages
import com.jetbrains.cidr.ArchitectureType
import com.jetbrains.cidr.execution.debugger.backend.DebuggerDriverConfiguration
import com.jetbrains.cidr.execution.debugger.backend.gdb.GDBDriverConfiguration
import com.jetbrains.cidr.execution.debugger.backend.lldb.LLDBDriverConfiguration
import java.io.File

class ZigDefaultDebuggerDriverConfigurationProvider: ZigDebuggerDriverConfigurationProvider() {
    override suspend fun getDebuggerConfiguration(project: Project, isElevated: Boolean, emulateTerminal: Boolean): DebuggerDriverConfiguration? {
        val settings = ZigDebuggerSettings.instance
        val service = zigDebuggerToolchainService
        val kind = settings.debuggerKind
        if (!availabilityCheck(project, kind))
            return null

        return when(val availability = service.debuggerAvailability(kind)) {
            DebuggerAvailability.Bundled -> when(kind) {
                DebuggerKind.LLDB -> ZigLLDBDriverConfiguration(isElevated, emulateTerminal)
                DebuggerKind.GDB -> ZigGDBDriverConfiguration(isElevated, emulateTerminal)
                DebuggerKind.MSVC -> throw AssertionError("MSVC is never bundled")
            }
            is DebuggerAvailability.Binaries -> when(val binary = availability.binaries) {
                is LLDBBinaries -> ZigCustomBinariesLLDBDriverConfiguration(binary, isElevated, emulateTerminal)
                is GDBBinaries -> ZigCustomBinariesGDBDriverConfiguration(binary, isElevated, emulateTerminal)
                is MSVCBinaries -> ZigMSVCDriverConfiguration(binary, isElevated, emulateTerminal)
                else -> throw AssertionError("Unreachable")
            }
            DebuggerAvailability.Unavailable,
            DebuggerAvailability.NeedToDownload,
            DebuggerAvailability.NeedToUpdate -> throw AssertionError("Unreachable")
        }
    }
}

private suspend fun availabilityCheck(project: Project, kind: DebuggerKind): Boolean {
    val service = zigDebuggerToolchainService
    val availability = service.debuggerAvailability(kind)
    val (message, action) = when(availability) {
        DebuggerAvailability.Unavailable -> return false
        DebuggerAvailability.NeedToDownload ->
            ZigDebugBundle.message("debugger.run.unavailable.reason.download") to ZigDebugBundle.message("debugger.run.unavailable.reason.download.button")
        DebuggerAvailability.NeedToUpdate ->
            ZigDebugBundle.message("debugger.run.unavailable.reason.update") to ZigDebugBundle.message("debugger.run.unavailable.reason.update.button")
        DebuggerAvailability.Bundled,
            is DebuggerAvailability.Binaries -> return true
    }

    val downloadDebugger = if (!ZigDebuggerSettings.instance.downloadAutomatically) {
        showDialog(project, message, action)
    } else {
        true
    }

    if (downloadDebugger) {
        val result = withEDTContext(ModalityState.any()) {
            service.downloadDebugger(project, kind)
        }
        if (result is ZigDebuggerToolchainService.DownloadResult.Ok) {
            return true
        }
    }
    return false
}

private suspend fun showDialog(project: Project, message: String, action: String): Boolean {
    val doNotAsk = object: DoNotAskOption.Adapter() {
        override fun rememberChoice(isSelected: Boolean, exitCode: Int) {
            if (exitCode == Messages.OK) {
                ZigDebuggerSettings.instance.downloadAutomatically = isSelected
            }
        }
    }

    return withEDTContext(ModalityState.any()) {
        MessageDialogBuilder
            .okCancel(ZigDebugBundle.message("debugger.run.unavailable"), message)
            .yesText(action)
            .icon(Messages.getErrorIcon())
            .doNotAsk(doNotAsk)
            .ask(project)
    }
}

private open class ZigLLDBDriverConfiguration(private val isElevated: Boolean, private val emulateTerminal: Boolean): LLDBDriverConfiguration() {
    override fun getDriverName() = "Zig LLDB"
    override fun isElevated() = isElevated
    override fun emulateTerminal() = emulateTerminal
}
private class ZigCustomBinariesLLDBDriverConfiguration(
    private val binaries: LLDBBinaries,
    isElevated: Boolean,
    emulateTerminal: Boolean
) : ZigLLDBDriverConfiguration(isElevated, emulateTerminal) {
    override fun useSTLRenderers() = false
    override fun getLLDBFrameworkFile(architectureType: ArchitectureType): File = binaries.frameworkFile.toFile()
    override fun getLLDBFrontendFile(architectureType: ArchitectureType): File = binaries.frontendFile.toFile()
}

private open class ZigGDBDriverConfiguration(private val isElevated: Boolean, private val emulateTerminal: Boolean): GDBDriverConfiguration() {
    override fun getDriverName() = "Zig GDB"
    override fun isAttachSupported() = false
    override fun isElevated() = isElevated
    override fun emulateTerminal() = emulateTerminal
}
private class ZigCustomBinariesGDBDriverConfiguration(
    private val binaries: GDBBinaries,
    isElevated: Boolean,
    emulateTerminal: Boolean
) : ZigGDBDriverConfiguration(isElevated, emulateTerminal) {
    override fun getGDBExecutablePath() = binaries.gdbFile.sanitizedPathString!!
}

private class ZigMSVCDriverConfiguration(
    private val binaries: MSVCBinaries,
    private val isElevated: Boolean,
    private val emulateTerminal: Boolean
): MSVCDriverConfiguration() {
    override val debuggerExecutable get() = binaries.msvcFile
    override fun getDriverName() = "Zig MSVC"
    override fun getConsoleLanguage() = ZigLanguage
    override fun isElevated() = isElevated
    override fun emulateTerminal() = emulateTerminal
}